pgex/examples/collision_shapes/main.py

```diff
@@ -0,0 +1,335 @@
+
+    for vertex in polygon.local_to_world():
+        projection = vertex.dot(axis)
+
+        if projection < min_: min_ = projection
+
+        if projection > max_: max_ = projection
+
+    return (min_, max_)
+
+def polygon_x_circle(polygon: Polygon, circle: Circle) -> bool:
+    """ Check collision between polygon and circle. """
+
+    vertices = polygon.local_to_world()
+    n = len(vertices)
+
+    for i in range(n):
+        va = vertices[i]
+        vb = vertices[(i + 1) % n]
+
+        edge = vb - va
+        axis = edge.normalize()
+        axis = pygame.Vector2(-axis.y, axis.x)
+
+        min_a, max_a = project_polygon(polygon, axis)
+        min_b, max_b = project_circle(circle, axis)
+
+        if min_a >= max_b or min_b >= max_a:
+            return False
+    
+    return True
+
+def polygon_x_rect(polygon: Polygon, rect: pygame.FRect) -> bool:
+    """ Check collision between polygon and rectangle. """
+
+    # IMPORTANT: A better approach would be clipping the polygon and testing
+    #            against the rect. But for now we are just treating the rect
+    #            as another polygon.
+
+    rect_center = pygame.Vector2(*rect.center)
+
+    rect_vertices = [
+        pygame.Vector2(*rect.topleft) - rect_center,
+        pygame.Vector2(*rect.topright) - rect_center,
+        pygame.Vector2(*rect.bottomright) - rect_center,
+        pygame.Vector2(*rect.bottomleft) - rect_center
+    ]
+
+    rect_poly = Polygon(rect_center, rect_vertices)
+
+    return polygon_x_polygon(polygon, rect_poly)
+
+def polygon_x_polygon(polygon_a: Polygon, polygon_b: Polygon) -> bool:
+    """ Check collision between two polygons. """
+
+    vertices_a = polygon_a.local_to_world()
+    vertices_b = polygon_b.local_to_world()
+    na = len(vertices_a)
+    nb = len(vertices_b)
+
+    # Check for polygon A's edges
+    for i in range(na):
+        va = vertices_a[i]
+        vb = vertices_a[(i + 1) % na]
+
+        edge = vb - va
+        axis = edge.normalize()
+        axis = pygame.Vector2(-axis.y, axis.x)
+
+        min_a, max_a = project_polygon(polygon_a, axis)
+        min_b, max_b = project_polygon(polygon_b, axis)
+
+        if min_a >= max_b or min_b >= max_a:
+            return False
+    
+    # Check for polygon B's edges
+    for i in range(nb):
+        va = vertices_b[i]
+        vb = vertices_b[(i + 1) % nb]
+
+        edge = vb - va
+        axis = edge.normalize()
+        axis = pygame.Vector2(-axis.y, axis.x)
+
+        min_a, max_a = project_polygon(polygon_a, axis)
+        min_b, max_b = project_polygon(polygon_b, axis)
+
+        if min_a >= max_b or min_b >= max_a:
+            return False
+        
+    return True
+
+
+async def main() -> None:
+    pygame.init()
+    window = pygame.display.set_mode((450, 140))
+    pygame.display.set_caption("PGEX Collision Shapes Example")
+    clock = pygame.time.Clock()
+    
+    # Prepare scene
+
+    circles = []
+
+    circles.append( Circle(pygame.Vector2(50, 50), 30) )
```

Comment:
Not sure if the extra blank spaces between parenthesis of `append` are good to have.